src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -734,7 +736,18 @@ public virtual int Run(TextWriter consoleOutput, CancellationToken cancellationT
         /// <param name="additionalTexts">Any additional texts that should be passed to the generators when run.</param>
         /// <param name="generatorDiagnostics">Any diagnostics that were produced during generation</param>
         /// <returns>A compilation that represents the original compilation with any additional, generated texts added to it.</returns>
-        private protected virtual Compilation RunGenerators(Compilation input, ParseOptions parseOptions, ImmutableArray<ISourceGenerator> generators, AnalyzerConfigOptionsProvider analyzerConfigOptionsProvider, ImmutableArray<AdditionalText> additionalTexts, DiagnosticBag generatorDiagnostics) { return input; }
+        private protected Compilation RunGenerators(Compilation input, ParseOptions parseOptions, ImmutableArray<ISourceGenerator> generators, AnalyzerConfigOptionsProvider analyzerConfigOptionsProvider, ImmutableArray<AdditionalText> additionalTexts, DiagnosticBag generatorDiagnostics)
+        {
+            var cacheKey = Arguments.GetOutputFilePath(GetOutputFileName(input, cancellationToken: default));
```

Comment:
I'm using the output file path and name as a cache key, as it seemed like a fairly good way of tracking 'is this the same project'. 

A little concerned about the case we don't have an output file name, as we actually use binding to figure it out. I wonder if in those cases we should just make a random key and effectively disable caching?

Reply:
A couple of thoughts:

* If there isn't an output file name, both C# and VB seem to consider this to be an error case, so disabling caching seems fine to me.
* VB asserts that this method will never be called unless there _is_ a file name, and that assert may no longer be true if we're calling it from here?

Reply:
I don't believe the second point was addressed?

Reply:
>If there isn't an output file name, both C# and VB seem to consider this to be an error case, 

Ah if it were so simple, C# actually uses the filename of the first source file passed that is not a script. 

Missed the point about VB. Given that I think we should just not cache in those cases. Will update + test.